### PR TITLE
feat(devops-1926): move promotion environments selection to `joy release promote` command

### PR DESCRIPTION
**Pull request by silphid**
feat: better syntax/usage help message
feat: can specify source/target promotion envs via flags
fix: all commands now properly determine catalog dir
fix: promotion preserves order of locked target nodes
fix: promotion no longer affect target value when source node is locked

### CODE REVIEW

**Comment by j-martin** [comment]
> feat: better syntax/usage help message
feat: can specify source/target promotion envs via flags
fix: all commands now properly determine catalog dir
fix: promotion preserves order of locked target nodes
fix: promotion no longer affect target value when source node is locked

As a general comment, those should really be individual commits. The current diff is > 1k lines, which is a big ask for reviewers. Would it be possible to split your change into distinct commits?

**Comment by silphid** [comment]
> As a general comment, those should really be individual commits. The current diff is > 1k lines, which is a big ask for reviewers. Would it be possible to split your change into distinct commits?

I fully understand and agree, ideally yes. However, most of the changes were inter-related / inter-dependent, or bugs that were discovered in the process of implementing the changes and that could not be addressed separately without doing the work twice (both with old and new approaches).  That is mostly due to this PR involving a major refactoring and change of approach.

I _could_ artificially break it down into smaller commits within same PR, just to help with review, but those commits would not be functional/buildable on their own, because of the interwoven nature of things here. If you really do think it's worth it, I could spend the extra effort to create those commits.  Actually, I had many commits that I squashed before creating the PR, because they just added confusion to the bigger picture, I think artificial commits would be more helpful to understand and review the PR.

**Comment by j-martin** [comment]
> [...] were discovered in the process of implementing the changes and that could not be addressed separately without doing the work twice (both with old and new approaches). That is mostly due to this PR involving a major refactoring and change of approach.

Yes, this always happens 😬. In such cases, I tend to backtrack and move the bug fixes and refactoring earlier in the git history. But yeah, even with proper tooling to work with `git rebase --interactive`, it can still be a pain.  

> I think artificial commits would be more helpful to understand and review the PR.

That would be great, a big improvement over this one huge commit. :)

**Review by j-martin — COMMENTED** [review]
I'll stop reviewing until the commits are split. Just posting those two nits for now. :)

**Review by j-martin — APPROVED** [review]
LGTM, just one nit (more of a question). Thanks for splitting the big commit. Next time, let's try to have a few PRs. :)

File: cmd/joy/release.go [hunk 1]
```diff
@@ -73,36 +73,76 @@ func NewReleasePromoteCmd() *cobra.Command {
...
 	cmd := &cobra.Command{
 		Use:     "select",
 		Aliases: []string{"sel"},
-		Short:   "Choose releases to work with",
-		Long: `Choose releases to work with.
+		Short:   "Select releases to include in listings and promotions",
+		Long: `Select releases to include in listings and promotions.
+
+Usage:
+  joy release select [flags]
```
[the remark this answers is not in the document]
Comment (silphid) [reply cmd/joy/release.go:142]: Cobra puts the `[flags]` token in all usage message in lower-case by default, I'm just sticking that when I override that usage, if that's ok for you, just to stay coherent with all the other commands.

File: cmd/joy/release.go [hunk 2]
```diff
@@ -114,20 +113,37 @@ func NewReleasePromoteCmd() *cobra.Command {
...
 
-Only selected releases will be included in releases table and during promotion.`,
+Flags:
+  -a, --all          Select all releases (non-interactive)
+`,
```
[the remark this answers is not in the document]
Comment (silphid) [reply cmd/joy/release.go:146]: Indeed, that one should have been left with default usage, not sure why I went that much out of my way here... 🤔

**Pull request merged**